### PR TITLE
For Mentor Review | concurrency | Farzaneh Haghani

### PR DESCRIPTION
**Pull request by farzaneh-haghani**
## Self checklist

- [x] I have titled my PR with `For Mentor Review | {Project Name} | {My Name}`
- [ ] I have included a list of changes
- [x] I have tested my changes
- [ ] I have explained my changes
- [x] I have asked someone to review my changes

### Change list

- List changes
- This might relate to specific commits in your PR but just pull out the key lines

### CODE REVIEW

File: projects/concurrency/implement-channel.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+			break
+		}
+	}
+}
+
+func (c *channel) Recv() string {
+	for {
+		valueToReturn := c.value.Load()
+		if c.value.Load() != nil {
+			if c.value.CompareAndSwap(valueToReturn, nil) {
```
Comment (illicitonion) [inline projects/concurrency/implement-channel.go]: This works, but could be simpler by doing a single operation on `c.value` rather than two that you're currently doing (`Load` + `CompareAndSwap`) - can you work out how?

File: projects/concurrency/main.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
+package main
+
+import "fmt"
+
+type Value[K comparable] struct {
+	data any
```
Comment (illicitonion) [inline projects/concurrency/main.go]: Rather than `any`, you may want to add a new generic type for the value as well, so that you don't need to make type assertions when retrieving the data.

File: projects/concurrency/main.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	m     map[K]Value[K]
+	first *K
+	last  *K
+}
+
+func NewCache[K comparable](entryLimit int) Cache[K] { //All K should be unique && same type && comparable type like primitive that can compare with each other
+	return Cache[K]{
+		size:  entryLimit,
+		m:     make(map[K]Value[K]),
+		first: nil,
```
Comment (illicitonion) [inline projects/concurrency/main.go]: I think your code would probably be simpler if you had two fields here, a `map` and a `LinkedList` (where you may need to implement `LinkedList` yourself, or may be able to use a library).

Right now, you have a linked list implementation in-line in your cache. While this can work, it has three big drawbacks:
1. It's harder to understand the code. Seeing `c.list.PushBack(value)` is easier to understand than updating `last` and also maybe updating `first` if the list is empty and maybe also updating `next` for some existing value and maybe also updating `prev` for some existing value. _Encapsulating_ the linked list in its own type means we can _abstract_ over it, and think in higher level terms. It's also easier to test the linked list on its own.
2. It's harder to re-use the linked list. If you've implemented a good linked list, you may want to use it for other projects. That's really easy to do if it's a standalone type, and really hard to do if it's mixed in with your cache implementation.
3. It's harder to use someone else's linked list implementation, or swap it with some other data structure. For instance, the Go standard library actually contains a linked list implementation (https://pkg.go.dev/container/list) which you could use if you wanted. If you write your own linked list separately, it's easy to replace just that one part with someone else's implementation. But it's going to be harder to disentangle the current implementation.

File: projects/concurrency/main.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		value.prev = c.last
+		value.next = nil
+		c.m[key] = value
+
+		c.last = &key
+		return false
+	}
+}
+
+func (c *Cache[K]) Get(key K) (*Value[K], bool) {
```
Comment (illicitonion) [inline projects/concurrency/main.go]: Get should also make the key be the most recently used. i.e. if you add 10 values, then get one, the other un-read 9 values should be evicted (if necessary) before the one gotten one.

File: projects/concurrency/main.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+func NewCache[K comparable](entryLimit int) Cache[K] { //All K should be unique && same type && comparable type like primitive that can compare with each other
+	return Cache[K]{
+		size:  entryLimit,
+		m:     make(map[K]Value[K]),
+		first: nil,
+		last:  nil,
+	}
+}
+
+func (c *Cache[K]) Put(key K, value Value[K]) bool {
```
Comment (illicitonion) [inline projects/concurrency/main.go]: Ideally the `Value` type wouldn't leak into your public interface. It's a useful implementation detail for you, but as far as a user is concerned they should ideally be able to just `c.Put("greeting", "hello")` without needing to know about your `Value` wrapping. Same with `Get` - it should return a pointer to a `string` or `any` or whatever, not a `Value`.

**Review by illicitonion — COMMENTED** [review]
This is generally looking reasonable for a non-threadsafe cache implementation, but I'd strongly recommend you switch out to a separate linked list type (which you can implement yourself or take from a library).

Right now there are several places where you dereference pointers which you don't know aren't nil, and you have a lot of conditional branches that are quite hard to follow - using a separate linked list would improve both of these situations.

**Comment by farzaneh-haghani** [comment]
> This is generally looking reasonable for a non-threadsafe cache implementation, but I'd strongly recommend you switch out to a separate linked list type (which you can implement yourself or take from a library).
> 
> Right now there are several places where you dereference pointers which you don't know aren't nil, and you have a lot of conditional branches that are quite hard to follow - using a separate linked list would improve both of these situations.

I changed it but I had to store a copy of key in each node to can delete from link list and map table. But before I didn't need it because each key in map was for value and next, prev pointers. I agree that I had mistakes in my implementation and was harder to understand but I think it could be improved by using small functions to use less memory.

File: projects/concurrency/main.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+		l.first = newNode
+	} else {
+		l.last.next = newNode
+		newNode.prev = l.last
+	}
+	l.last = newNode
+	return newNode
+}
+
+func (l *List[K, V]) moveNode(currentNode *Node[K, V]) {
```
Comment (illicitonion) [inline projects/concurrency/main.go]: Where does this move the node to? Can you think of a name for this function which makes it more clear?

File: projects/concurrency/main.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+	if l.last != currentNode {
```
Comment (illicitonion) [inline projects/concurrency/main.go]: It would probably be slightly easier to follow this by writing:
```go
if l.last == currentNode {
  return
}
// Rest of implementation goes here
```

This has a few advantages - the main one is that I don't need to remember I'm in an if (and be thinking about what the else may be at the end). "If I'm the last node, there's nothing to do. Otherwise, do the rest of the function." - this framing allows me to stop thinking "is there an else" as soon as I see the `return`.

File: projects/concurrency/main.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+			l.first = currentNode.next
+			l.first.prev = nil
+		}
+		currentNode.prev = l.last
+		l.last = currentNode
+		currentNode.next = nil
+	}
+}
+
+func (l *List[K, V]) deleteNode() (K, int) {
```
Comment (illicitonion) [inline projects/concurrency/main.go]: This name doesn't say _which_ node is deleted. Is it the first? The last? A random one?

Can you think of a name for this function which makes clear which node is getting deleted?

File: projects/concurrency/main.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+	deleted := l.first.key
+	entriesRead := l.first.entriesRead
+	if l.first == l.last {
+		l.first = nil
+		l.last = nil
+	} else {
+		l.first = l.first.next
+		l.first.prev = nil
+	}
+	return deleted, entriesRead
```
Comment (illicitonion) [inline projects/concurrency/main.go]: It's fine that your `List` implementation knows about `entriesRead`, but it makes this code less generic and re-usable. Right now your `List` only makes sense to be used inside your cache, but with a small modification you could also use it anywhere else you need a linked list.

You may want to introduce an `Entry` type like:

```go
type Entry[K any, V any] struct {
    key K
    value V
    entriesRead int 
}
```

so that you can just have a `List[Entry]`. This function could return an `Entry`, and then your `Cache` type can get the fields out of that returned value.

File: projects/concurrency/main.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func (c *Cache[K, V]) Put(key K, value V) bool {
+	c.rwm.Lock()
+	if currentNode, isExisted := c.m[key]; isExisted {
+		currentNode.value = value
+		c.l.moveNode(currentNode)
+		c.rwm.Unlock()
```
Comment (illicitonion) [inline projects/concurrency/main.go]: If you always only unlock the lock just before returning, we tend to `defer c.rwm.Unlock()` right after locking it. 

This serves two purposes:
1. It means we'll definitely unlock it and can't forget, even if we change the code
2. It means that when I'm reading the code, I know "this whole function is guarded by a lock", and don't need to pay as much attention to whether the locking is correct / we forgot to unlock somewhere / we maybe double unlock, etc

File: projects/concurrency/main.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+	newNode := c.l.addNode(key, value)
+	c.m[key] = newNode
+	c.s.entriesNeverRead++
+	c.s.writesCount++
+	c.rwm.Unlock()
+	return false
+}
+
+func (c *Cache[K, V]) Get(key K) (*V, bool) {
+	c.rwm.RLock()
```
Comment (illicitonion) [inline projects/concurrency/main.go]: How is calling `RLock()` different to calling `Lock()`? Why is is safe to call `RLock()` here instead of `Lock()`? Why is it better than calling `Lock()`?

File: projects/concurrency/main.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+		}
+		c.rwm.RUnlock()
+		return &currentNode.value, true
+	}
+	c.s.unreadCount++
+	c.rwm.RUnlock()
+	return nil, false
+}
+
+var wg sync.WaitGroup
```
Comment (illicitonion) [inline projects/concurrency/main.go]: This should probably be declared inside `main()` rather than as a static global :)

**Review by illicitonion — COMMENTED** [review]
This is generally looking really good!

> I changed it but I had to store a copy of key in each node to can delete from link list and map table. But before I didn't need it because each key in map was for value and next, prev pointers. I agree that I had mistakes in my implementation and was harder to understand but I think it could be improved by using small functions to use less memory.

This is really interesting, let's talk about it!

It is often the case that the clearest code is in some way more expensive (slower, uses more memory, makes more network calls, ...) than the cheapest code. Maybe you could try make the best "everything in one struct" implementation you can, and we can compare this one and that one?

File: projects/concurrency/list/list.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func NewList[K comparable, V any]() *List[K, V] {
+	return &List[K, V]{
+		first: nil,
+		last:  nil,
+	}
+}
+
+func (l *List[K, V]) AddNode(key K, value V) *Node[K, V] {
```
Comment (illicitonion) [inline projects/concurrency/list/list.go]: I'd probably rename this to say _where_ the node is added.

File: projects/concurrency/list/list_test.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+		require.Equal(t, deleted, 2)
+		require.NotEqual(t, l.first, newNode2)
+		require.Equal(t, l.first, newNode1)
+		require.Equal(t, l.first, l.last)
+		require.Equal(t, newNode1.next, newNode1.prev)
+
+		deleted2, _ := l.DeleteFirstNode()
+		require.Equal(t, deleted2, 1)
+		require.NotEqual(t, l.first, newNode1)
+		require.Equal(t, l.first, l.last)
```
Comment (illicitonion) [inline projects/concurrency/list/list_test.go:48]: I would probably also explicitly required that they're both `nil`.

**Review by illicitonion — COMMENTED** [review]
Looking great!

**Comment by farzaneh-haghani** [comment]
> Looking great!

Thank you. I committed the custom cache after your review but I mixed up the test :)

File: projects/concurrency/cache/cache.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+func (c *Cache[K, V]) PrintStatics(w io.Writer, static Static, length int) {
+	totalHit := static.ReadCount + static.UnreadCount
+	hitRate := float32(static.ReadCount) / float32(totalHit) * 100
+	if hitRate == 0 {
+		io.WriteString(w, "Never have had any read from this cache!")
+	} else {
+		hit := fmt.Sprintf("Hit rate: %.2f\n", hitRate)
+		io.WriteString(w, hit)
```
Comment (illicitonion) [inline projects/concurrency/cache/cache.go]: This is fine, but FYI there is a function `fmt.Fprintf` which allows you to do both of these calls in one :)

File: projects/concurrency/computing-cache/computing-cache.go [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+	}
+	if len(c.M) >= c.Size {
+		deleted, entriesRead := c.L.DeleteFirstNode()
+		delete(c.M, deleted)
+		c.S.TotalReadExisted -= entriesRead
+	}
+	value := c.Creator(key)
+	newNode := c.L.AddNodeToLast(key, value)
+	c.M[key] = newNode
+	return value
```
Comment (illicitonion) [inline projects/concurrency/computing-cache/computing-cache.go]: This feels like it's really digging into the internals of your cache (and you had to make the field `L` public to be able to do so). Things like the `L` field, the idea that we have a `map` and a `List` internally, and the relationship between them are very much internal implementation details of our cache type, not part of its public API.

It's interesting to use your existing cache type as a building block here - if you're going to do so, though, you should only use the _public_ API of that type - so you'd instead do something like:

```go
value := c.Creator(key);
c.cache.Put(key, value)
```

In particular, you have a problem here that there's no locking that happens between you deleting an old entry, and you creating a new one. There are ways you can add that locking, but let's skip over thinking about that for now :)

The big challenge with a computing cache is that if two `Get` calls come in at the same time, you only want to do the computing once. Let's imagine we had a `Creator` function which sleeps for 5 seconds and then returns a value, and we have two goroutines trying to get the value for the first time in parallel. We only want to call `Creator` once, not twice.

File: projects/concurrency/list/list.go [hunk 17]
```diff
@@ -0,0 +1,86 @@
+package list
+
+type Entry[K comparable, V any] struct {
+	key         K
+	Value       V
+	IsRead      bool //Question 2
+	EntriesRead int  //Question 3
+}
+type Node[K comparable, V any] struct {
+	Data Entry[K, V]
```
Comment (illicitonion) [inline projects/concurrency/list/list.go:10]: Is "Has this value been read" a general concept of all lists? I'd argue not - it's a caching concept, not a list concept.

Which suggests our `List` type shouldn't know about the `Entry` type - our `List` should just have generic `V`s, and our `Cache` type should make a `List` where the values are `Entry` values (the type of which should be defined in the cache package not the list one).
